### PR TITLE
support python 3.9

### PR DESCRIPTION
**Pull request by Borda**
## What does this PR do?

includes:
- testing / CI for 3.9

## Before submitting
- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? 
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:  

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects).
 

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by Borda** [comment]
missing wheels for pytorch - https://github.com/pytorch/pytorch/issues/45731

**Comment by tchaton** [comment]
@Borda Should we close this PR ?

**Comment by Borda** [comment]
> @Borda Should we close this PR ?

Let's keep it as a draft bc it is ready, just need PT to release wheels...

**Comment by ydcjeff** [comment]
we might need to wait a bit as torchvision and torchtext doesn't have 3.9 build in nightly.

**Comment by Borda** [comment]
blocked by https://github.com/microsoft/onnxruntime/issues/6125

**Comment by Borda** [comment]
Consider building own onnxruntime https://github.com/microsoft/onnxruntime/issues/6125#issuecomment-767067399

**Review by awaelchli — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM !

File: .github/workflows/ci_test-full.yml [hunk 1]
```diff
@@ -15,15 +15,18 @@ jobs:
...
         exclude:
           # # todo: segmentation fault for minimal and hanging for latest
           - python-version: 3.8
             os: ubuntu-18.04
 
     # Timeout: https://stackoverflow.com/a/59076067/4521646
-    timeout-minutes: 35  # TODO: the macOS is taking too long, probably caching did not work...
+    # TODO: the macOS is taking too long, probably caching did not work...
+    # todo: building own onnxruntime takes very long...
+    timeout-minutes: 75
```
Comment (Borda) [inline .github/workflows/ci_test-full.yml:28]:
```suggestion
    timeout-minutes: 40
```

File: .github/workflows/ci_test-full.yml [hunk 2]
```diff
@@ -97,18 +108,27 @@ jobs:
...
 
+    - name: Build from source for Python 3.9
+      if: matrix.python-version == 3.9
+      run: |
+        pip install "pip>=21.0" -U  # needed for later instalation onnxruntime
+        python -c "import re ; fname='requirements/extra.txt' ; req = re.sub(r'onnx>=[\d\.]+', '', open(fname).read()) ; open(fname, 'w').write(req)"
+        python -c "import re ; fname='requirements/extra.txt' ; req = re.sub(r'onnxruntime>=[\d\.]+', '', open(fname).read()) ; open(fname, 'w').write(req)"
+        bash requirements/install_ONNX.sh
+      shell: bash
+
```
Comment (Borda) [inline .github/workflows/ci_test-full.yml:119]:
```suggestion
```

File: .github/workflows/ci_test-full.yml [hunk 3]
```diff
@@ -97,18 +108,27 @@ jobs:
...
         HOROVOD_WITHOUT_MXNET: 1
         HOROVOD_WITHOUT_TENSORFLOW: 1
       run: |
-        # python -m pip install --upgrade --user pip
-        pip install --requirement requirements.txt --find-links https://download.pytorch.org/whl/cpu/torch_stable.html --quiet --upgrade
-        pip install --requirement ./requirements/devel.txt --find-links https://download.pytorch.org/whl/cpu/torch_stable.html --quiet --upgrade
         python --version
         pip --version
+        # python -m pip install --upgrade --user pip
+        pip install --requirement requirements.txt --quiet --upgrade --find-links https://download.pytorch.org/whl/cpu/torch_stable.html
```
Comment (Borda) [inline .github/workflows/ci_test-full.yml:130]:
```suggestion
```

File: .github/workflows/ci_test-full.yml [hunk 4]
```diff
@@ -52,6 +54,15 @@ jobs:
         open(fname, 'w').writelines(lines)
       shell: python
 
+    - name: Adjust minimal for Python 3.9
+      if: matrix.requires == 'minimal' && matrix.python-version == 3.9
+      run: |
+        import re
+        fname='requirements.txt'
+        req = re.sub(r'torch>=[\d\.]+', 'torch>=1.7.1', open(fname).read())
+        open(fname, 'w').write(req)"
```
Comment (Borda) [inline .github/workflows/ci_test-full.yml:63]:
```suggestion
        open(fname, 'w').write(req)
```

**Comment by Borda** [comment]
missing ONNX wheel https://github.com/onnx/onnx/issues/3329

**Review by s-rog — APPROVED** [review]

**Comment by Borda** [comment]
@tgaddair mind have look at CPU failing test for Linux and py3.9?

**Comment by tgaddair** [comment]
Hey @Borda, certainly, I'll take a look.

**Comment by tgaddair** [comment]
@Borda, at first glance, it looks like there is a race condition related to using `coverage` in parallel with Horovod / MPI (see: https://github.com/nedbat/coveragepy/issues/883). I would suggest disabling it in the Horovod test for now to see if that resolves the issue.

**Comment by Borda** [comment]
@tgaddair I have updated PR with master and seems like last failing test are Horovod, mind have look?
```
FAILED tests/models/test_horovod.py::test_horovod_cpu - assert 1 == 0
FAILED tests/models/test_horovod.py::test_horovod_cpu_implicit - assert 1 == 0
```

File: .github/workflows/ci_dockers.yml [hunk 5]
```diff
@@ -109,8 +109,10 @@ jobs:
             pytorch_version: 1.4
           - python_version: 3.7
             pytorch_version: 1.7
-          - python_version: 3.7
+          - python_version: 3.8
             pytorch_version: 1.8
+          #  - python_version: 3.9
```
Comment (justusschock) [inline .github/workflows/ci_dockers.yml:114]: why not testing with it then?

**Review by justusschock — APPROVED** [review]

**Review by kaushikb11 — APPROVED** [review]
🍾

File: .github/workflows/ci_dockers.yml (hunk 5 above)
Comment (Borda) [reply .github/workflows/ci_dockers.yml:114]: there was some problem in past, so I would rather leave it for separate PR

**Review by edenlightning — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]

**Pull request merged**